{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘tidyverse’ was built under R version 4.2.2”\n",
      "Warning message:\n",
      "“package ‘ggplot2’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘tibble’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘tidyr’ was built under R version 4.2.2”\n",
      "Warning message:\n",
      "“package ‘readr’ was built under R version 4.2.2”\n",
      "Warning message:\n",
      "“package ‘purrr’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘dplyr’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘stringr’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘forcats’ was built under R version 4.2.2”\n",
      "Warning message:\n",
      "“package ‘lubridate’ was built under R version 4.2.2”\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Warning message:\n",
      "“package ‘DESeq2’ was built under R version 4.2.3”\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘BiocGenerics’ was built under R version 4.2.1”\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.2.3”\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomicRanges’ was built under R version 4.2.3”\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomeInfoDb’ was built under R version 4.2.3”\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Warning message:\n",
      "“package ‘SummarizedExperiment’ was built under R version 4.2.2”\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘MatrixGenerics’ was built under R version 4.2.1”\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Warning message:\n",
      "“package ‘matrixStats’ was built under R version 4.2.3”\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.2.3”\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘sva’ was built under R version 4.2.3”\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "\n",
      "Attaching package: ‘nlme’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "This is mgcv 1.9-0. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "\n",
      "Attaching package: ‘genefilter’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    spec\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Warning message:\n",
      "“package ‘gridExtra’ was built under R version 4.2.3”\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘patchwork’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:genefilter’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘umap’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘ggsci’ was built under R version 4.2.3”\n",
      "Warning message:\n",
      "“package ‘viridis’ was built under R version 4.2.3”\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Warning message:\n",
      "“package ‘viridisLite’ was built under R version 4.2.3”\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(DESeq2)\n",
    "library(sva)\n",
    "\n",
    "source(\"../../utils/plots_eda.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] ‘1.38.0’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "[1] ‘3.46.0’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# version of DESeq2\n",
    "packageVersion(\"DESeq2\")\n",
    "\n",
    "# version of sva\n",
    "packageVersion(\"sva\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "datasets <- c(\"GSE129508\", \"GSE58135\", \"GSE149276\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Removing sample GSM3714613 from GSE129508\"\n",
      "[1] \"Samples: 25; Features: 35238\"\n",
      "[1] \"Normalized Samples: 25; Features: 35238\"\n",
      "[1] \"..plotting..\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22m`aes_string()` was deprecated in ggplot2 3.0.0.\n",
      "\u001b[36mℹ\u001b[39m Please use tidy evaluation idioms with `aes()`.\n",
      "\u001b[36mℹ\u001b[39m See also `vignette(\"ggplot2-in-packages\")` for more information.”\n",
      "No id variables; using all as measure variables\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Combined Samples: 25; Features: 35238\"\n",
      "[1] \" \"\n",
      "[1] \"Samples: 75; Features: 35238\"\n",
      "[1] \"Normalized Samples: 75; Features: 35238\"\n",
      "[1] \"..plotting..\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No id variables; using all as measure variables\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Combined Samples: 100; Features: 35238\"\n",
      "[1] \" \"\n",
      "[1] \"Samples: 31; Features: 35238\"\n",
      "[1] \"Normalized Samples: 31; Features: 35238\"\n",
      "[1] \"..plotting..\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No id variables; using all as measure variables\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Combined Samples: 131; Features: 35238\"\n",
      "[1] \" \"\n",
      "[1] \"Plotting combined data\"\n",
      "[1] \"..plotting..\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No id variables; using all as measure variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_expression <- NULL\n",
    "all_metadata <- NULL\n",
    "\n",
    "for(dataset in datasets){\n",
    "    metadata <- read.table(paste0(\"before/\", dataset, \"/\", dataset, \".sample_info.tsv\"), header = TRUE, sep = \"\\t\")\n",
    "    expr_data <- read.table(paste0(\"before/\", dataset, \"/\", dataset, \".counts.tsv\"), header = TRUE, sep = \"\\t\")\n",
    "\n",
    "\n",
    "    # if sample GCSM3714613 is in expression or metadata, remove it\n",
    "    if(\"GSM3714613\" %in% colnames(expr_data)){\n",
    "        print(paste0(\"Removing sample GSM3714613 from \", dataset))\n",
    "        expr_data <- expr_data[, -which(colnames(expr_data) == \"GSM3714613\")]\n",
    "        metadata <- metadata[metadata$sample_id != \"GSM3714613\",]\n",
    "    }\n",
    "    \n",
    "    metadata$Dataset <- dataset\n",
    "    metadata <- metadata %>% \n",
    "        mutate(Status = ifelse(Basal == 1, \"Basal\", \"Luminal\")) %>%\n",
    "            # ifelse(is_LumA == 1, \"LumA\", \"LumB\"))) %>%\n",
    "        mutate(Status = as.factor(Status))\n",
    "\n",
    "    print(paste0(\"Samples: \", nrow(metadata), \"; Features: \", nrow(expr_data)))\n",
    "\n",
    "    #########################################################################################################################\n",
    "    # Normalization # \n",
    "    # Create a DESeq2 dataset object\n",
    "    expr_data <- expr_data %>% column_to_rownames(\"gene_id\")\n",
    "\n",
    "    dds <- DESeqDataSetFromMatrix(countData = expr_data, colData = metadata, design = ~ Status)\n",
    "    # Normalize using median of ratios method\n",
    "    dds <- estimateSizeFactors(dds)\n",
    "    normalized_counts <- counts(dds, normalized = TRUE)\n",
    "    expr_data <- as.data.frame(normalized_counts) %>% \n",
    "        rownames_to_column(\"gene_id\")\n",
    "    print(paste0(\"Normalized Samples: \", nrow(metadata), \"; Features: \", nrow(expr_data)))\n",
    "\n",
    "    #########################################################################################################################\n",
    "    #  plot\n",
    "    plot_res <- plot_diagnostic(expr_data, metadata, dataset,\n",
    "                                log_transform = FALSE, with_rowname = FALSE)\n",
    "    layout <- (plot_res[[1]] + plot_res[[2]] ) / \n",
    "              (plot_res[[3]] )\n",
    "    ggsave(paste0(\"before/\", dataset, \"/diagnostic_plot.png\"), \n",
    "                plot = layout, width = 12, height = 12)\n",
    "\n",
    "    # save data\n",
    "    if(is.null(all_metadata)){\n",
    "        all_metadata <- metadata\n",
    "        all_expression <- expr_data\n",
    "    } else {        \n",
    "        all_metadata <- rbind(all_metadata, metadata)\n",
    "        all_expression <- full_join(all_expression, expr_data, by = \"gene_id\")\n",
    "    }\n",
    "    print(paste0(\"Combined Samples: \", nrow(all_metadata), \"; Features: \", nrow(all_expression)))\n",
    "    print(\" \")\n",
    "}\n",
    "\n",
    "# plot the combined data\n",
    "print(\"Plotting combined data\")\n",
    "plot_res <- plot_diagnostic(all_expression, all_metadata, \"Combined\")\n",
    "layout <- (plot_res[[1]] + plot_res[[2]] ) / \n",
    "          (plot_res[[3]] )\n",
    "ggsave(\"before/diagnostic_plot.png\", \n",
    "            plot = layout, width = 12, height = 12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'Dataset'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Dataset</th><th scope=col>Status</th><th scope=col>n()</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GSE129508</td><td>Basal  </td><td> 4</td></tr>\n",
       "\t<tr><td>GSE129508</td><td>Luminal</td><td>21</td></tr>\n",
       "\t<tr><td>GSE149276</td><td>Basal  </td><td>17</td></tr>\n",
       "\t<tr><td>GSE149276</td><td>Luminal</td><td>14</td></tr>\n",
       "\t<tr><td>GSE58135 </td><td>Basal  </td><td>36</td></tr>\n",
       "\t<tr><td>GSE58135 </td><td>Luminal</td><td>39</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 6 × 3\n",
       "\\begin{tabular}{lll}\n",
       " Dataset & Status & n()\\\\\n",
       " <chr> & <fct> & <int>\\\\\n",
       "\\hline\n",
       "\t GSE129508 & Basal   &  4\\\\\n",
       "\t GSE129508 & Luminal & 21\\\\\n",
       "\t GSE149276 & Basal   & 17\\\\\n",
       "\t GSE149276 & Luminal & 14\\\\\n",
       "\t GSE58135  & Basal   & 36\\\\\n",
       "\t GSE58135  & Luminal & 39\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 6 × 3\n",
       "\n",
       "| Dataset &lt;chr&gt; | Status &lt;fct&gt; | n() &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| GSE129508 | Basal   |  4 |\n",
       "| GSE129508 | Luminal | 21 |\n",
       "| GSE149276 | Basal   | 17 |\n",
       "| GSE149276 | Luminal | 14 |\n",
       "| GSE58135  | Basal   | 36 |\n",
       "| GSE58135  | Luminal | 39 |\n",
       "\n"
      ],
      "text/plain": [
       "  Dataset   Status  n()\n",
       "1 GSE129508 Basal    4 \n",
       "2 GSE129508 Luminal 21 \n",
       "3 GSE149276 Basal   17 \n",
       "4 GSE149276 Luminal 14 \n",
       "5 GSE58135  Basal   36 \n",
       "6 GSE58135  Luminal 39 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_metadata %>%\n",
    "    group_by(Dataset, Status) %>% summarise(n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Dataset</th><th scope=col>n()</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GSE129508</td><td>25</td></tr>\n",
       "\t<tr><td>GSE149276</td><td>31</td></tr>\n",
       "\t<tr><td>GSE58135 </td><td>75</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " Dataset & n()\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t GSE129508 & 25\\\\\n",
       "\t GSE149276 & 31\\\\\n",
       "\t GSE58135  & 75\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 3 × 2\n",
       "\n",
       "| Dataset &lt;chr&gt; | n() &lt;int&gt; |\n",
       "|---|---|\n",
       "| GSE129508 | 25 |\n",
       "| GSE149276 | 31 |\n",
       "| GSE58135  | 75 |\n",
       "\n"
      ],
      "text/plain": [
       "  Dataset   n()\n",
       "1 GSE129508 25 \n",
       "2 GSE149276 31 \n",
       "3 GSE58135  75 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_metadata %>%\n",
    "    group_by(Dataset) %>% summarise(n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of features before filtering: 35238\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of features after filtering: 35238\"\n",
      "[1] \"Number of samples: 131\"\n"
     ]
    }
   ],
   "source": [
    "print(paste0(\"Number of features before filtering: \", nrow(all_expression)))\n",
    "all_expression <- na.omit(all_expression)\n",
    "print(paste0(\"Number of features after filtering: \", nrow(all_expression)))\n",
    "print(paste0(\"Number of samples: \", nrow(all_metadata)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Number of samples: 131\"\n",
      "[1] \"Number of features: 35238; Number of samples: 131\"\n"
     ]
    }
   ],
   "source": [
    "all_expression <- all_expression %>% column_to_rownames(\"gene_id\")\n",
    "all_expression <- all_expression[, all_metadata$sample_id]\n",
    "all_expression <- log2(all_expression + 1)\n",
    "print(paste0(\"Number of samples: \", nrow(all_metadata)))\n",
    "print(paste0(\"Number of features: \", nrow(all_expression), \"; Number of samples: \", ncol(all_expression)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 6415 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found3batches\n",
      "\n",
      "Adjusting for1covariate(s) or covariate level(s)\n",
      "\n",
      "Standardizing Data across genes\n",
      "\n",
      "Fitting L/S model and finding priors\n",
      "\n",
      "Finding parametric adjustments\n",
      "\n",
      "Adjusting the Data\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Plotting combined corrected data\"\n",
      "[1] \"..plotting..\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No id variables; using all as measure variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "design <- model.matrix(~all_metadata$Status)\n",
    "\n",
    "corrected_expr <- sva::ComBat(dat = all_expression, \n",
    "                              batch = all_metadata$Dataset, \n",
    "                              mod = design)\n",
    "\n",
    "corrected_expr <- as.data.frame(corrected_expr)\n",
    "\n",
    "# plot the combined corrected data\n",
    "print(\"Plotting combined corrected data\")\n",
    "plot_res <- plot_diagnostic(corrected_expr, all_metadata, \"Combined Corrected\",\n",
    "                            log_transform = TRUE, with_rowname = TRUE)\n",
    "layout <- (plot_res[[1]] + plot_res[[2]] ) / \n",
    "          (plot_res[[3]] )\n",
    "ggsave(\"after/diagnostic_plot_corrected.png\", \n",
    "            plot = layout, width = 12, height = 12)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save data for correction and after correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Save data prior to batch correction for GSE129508\"\n",
      "[1] \"Save data prior to batch correction for GSE58135\"\n",
      "[1] \"Save data prior to batch correction for GSE149276\"\n"
     ]
    }
   ],
   "source": [
    "all_metadata$batch <- as.numeric(as.factor(all_metadata$Dataset)) - 1\n",
    "# all_metadata$batch <- 0\n",
    "all_metadata$lum <- as.numeric(as.factor(all_metadata$Status))\n",
    "all_metadata$lum = all_metadata$lum - 1\n",
    "\n",
    "for (dataset in unique(all_metadata$Dataset)) {\n",
    "    print(paste0(\"Save data prior to batch correction for \", dataset))\n",
    "    dataset_metadata <- all_metadata[all_metadata$Dataset == dataset,]\n",
    "    dataset_metadata <- dataset_metadata %>% select(sample_id, lum, batch)\n",
    "    \n",
    "    dataset_expression <- all_expression[, dataset_metadata$sample_id]\n",
    "    dataset_expression <- dataset_expression %>% rownames_to_column(\"gene_id\")\n",
    "    \n",
    "    write.table(dataset_metadata, \n",
    "        file = paste0(\"before/\", dataset, \"/design.tsv\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "    write.table(dataset_expression, \n",
    "        file = paste0(\"before/\", dataset, \"/expr_for_correction.tsv\"), sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "\n",
    "}\n",
    "\n",
    "write.table(all_metadata %>% select(sample_id, lum, batch), \n",
    "    file = \"before/all_design.tsv\", sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "write.table(all_expression %>% rownames_to_column(\"gene_id\"),\n",
    "    file = \"before/all_expr_for_correction.tsv\", sep = \"\\t\", quote = FALSE, row.names = FALSE)\n",
    "write.table(corrected_expr %>% rownames_to_column(\"gene_id\"),\n",
    "    file = \"after/all_corrected_R_expr.tsv\", sep = \"\\t\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fed Correction ...\n",
    "\n",
    "... after FC app ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Samples: 26; Features: 35238\"\n",
      "[1] \"Combined Samples: 26; Features: 35238\"\n",
      "[1] \" \"\n",
      "[1] \"Samples: 32; Features: 35238\"\n",
      "[1] \"Combined Samples: 57; Features: 35238\"\n",
      "[1] \" \"\n",
      "[1] \"Samples: 76; Features: 35238\"\n",
      "[1] \"Combined Samples: 132; Features: 35238\"\n",
      "[1] \" \"\n"
     ]
    }
   ],
   "source": [
    "fed_expression <- NULL\n",
    "\n",
    "path_to_fed <- \"/home/yuliya/repos/cosybio/FedComBat/datasets/Breast_cancer_RNASeq/after/fed_res/\"\n",
    "\n",
    "for(i in 0:2){\n",
    "    expr_data <- read.table(paste0(path_to_fed, i, \"_bayes_data.csv\"), header = TRUE, sep = \",\")\n",
    "    print(paste0(\"Samples: \", ncol(expr_data), \"; Features: \", nrow(expr_data)))\n",
    "    # save data\n",
    "    if(is.null(fed_expression)){\n",
    "        fed_expression <- expr_data\n",
    "    } else {        \n",
    "        fed_expression <- full_join(fed_expression, expr_data, by = \"gene_id\")\n",
    "    }\n",
    "    print(paste0(\"Combined Samples: \", ncol(fed_expression), \"; Features: \", nrow(fed_expression)))\n",
    "    print(\" \")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Plotting combined data\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"..plotting..\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No id variables; using all as measure variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "fed_expression <- fed_expression %>% column_to_rownames(\"gene_id\") \n",
    "fed_expression <- fed_expression[, all_metadata$sample_id]\n",
    "\n",
    "# plot the combined data\n",
    "print(\"Plotting combined data\")\n",
    "plot_res <- plot_diagnostic(fed_expression, all_metadata, \"FedCombat corrected\",\n",
    "                            log_transform = TRUE, with_rowname = TRUE)\n",
    "layout <- (plot_res[[1]] + plot_res[[2]] ) / \n",
    "          (plot_res[[3]] )\n",
    "ggsave(\"after/diagnostic_plot_correcter_Fed.png\", \n",
    "            plot = layout, width = 12, height = 12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Min Max and mean absolute error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "corrected_expr <- as.data.frame(corrected_expr)\n",
    "fed_expression <- fed_expression[rownames(corrected_expr), colnames(corrected_expr)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Min difference: 0\"\n",
      "[1] \"Mean difference: 0.38892337828645\"\n",
      "[1] \"Max difference: 3.58996554293112\"\n"
     ]
    }
   ],
   "source": [
    "# Calculate value-to-value mean, max, and mean absolute difference\n",
    "mean_diff <- mean(as.matrix(abs(corrected_expr - fed_expression), na.rm = TRUE))\n",
    "max_diff <- max(abs(corrected_expr - fed_expression), na.rm = TRUE)\n",
    "min_diff <- min(abs(corrected_expr - fed_expression), na.rm = TRUE)\n",
    "\n",
    "print(paste0(\"Min difference: \", min_diff))\n",
    "print(paste0(\"Mean difference: \", mean_diff))\n",
    "print(paste0(\"Max difference: \", max_diff))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC/VBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///+NBy6HAAAACXBIWXMAABJ0AAASdAHeZh94AAAe5klEQVR4nO3debxUdf3H8c+9l5m7sCOCIIIg7poLpqKiIvzEJESlRIVkccdyKzFL0SwtJW3RLC2tCC13/alUIC5ILrggij/xRqaoiCjgxs79Pn7fc2Y7M3NHHe97Bu7l9fxj5sw5Z77zvcfzYpY7iDkATWYbewJAS0BIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEtLHcYVZd/L0e7N+h7XYrP3fM1G6fvTt0CKkpbrKEVlv0v3RJkff9UiE9Uxk83MdZD1/RYYdRt63LGjO1W/buKCFCaopUSIEOs4u7b4GQ3q2y/yt8pzPNut45fX3+w2/7eLCi/tprfx3dLXt3lBAhNUVwJp8yYcKE0Tv7hV4birpvgZB+ZZ8V0pFmp2c//JHDhx1Q56+r7mlkt+zdUUKE1BTBmbw8XDrFLz1b1H0LhHTgZ4Y0yOz8/Idf/asas9Zv5O+WvTtKiJCaIhPSo37p4WBhxU/26xDrcvjNwZuW6RVmM/31sq5mZ7vfme3s/nZAuzYDHwt2TIcUvcfQxAu1szMPEd16WvJ13Md5D/+gf6TTUmOmdjsxs3v96X1r2n71F2v9rn4aA9af0bbK5a49yD0+uEPdgBmJsd86d+e6ml0u/CC8EdnRrbtxYOdWW/a7YmmpDmrzREhNkTmT7/RLb/rruT2SZ++By/ytc836rnJuvNluq9xUs60nJz6bmOYyIWXdIy+krK2FQ3LDzLZoKBjSvbWJpYF+Mn4ae1zjXwq6rLW3mu3+j3hwqyosaWb7xLZuL7vsHdcemhy162c8cW6GCKkpUmfyR4/0NTvWLyzz533v3947sZXZcH9z9e5m33ePVFj1i87d7s/y6lPuvKKNWc916ZCy7zH/Pj/i1Fn/ST1A9tZ/z9rHbOSsWRtyHt67xS/OT46Z2u211O4LW5tdsGDOIcFkgmls2zO2Rx+Xtdbfs/u2/S4a4ofZ3w/3Tief7h1/2dNsh3XZO95gttNts6cdY3ZwWQ/1po6QmiLysVn8pE/9iklmbRb56ynJt0zzqq3V0zuYXePCc9XGu/APf5uWDinnHost6z1S7niNv0fyngpfWqae5XLeI53ln0z81dI21u7TcBrbL2ps7SHrnfPPYpX+CW+i2VafOPeefyq6K3vHMWa/8DfWjf72T4v7cKWFI6SmiIS0z5XB2wn/DPStYMP6jmaXBQv+RVSd2eAGlwgpiGFdW7NJ6ZBy7pETUu54BUN6yS/eXygk/3R58SrvYLPp4TRudY2tDd7OzfDXrzm3q9kZwT6zp02rz97Rv1rt+ae3SnM4mzNCaorUx9+nDe/g3zTMdg1VZpPDLf3NTgiuGwb7XTq9HSwG5+qaYGEvs7GpkHLvkR1S3ngFQ5rlF58uENKGykzvvwin8a5rbO1Hfu1r/vql8HGvTT1K9o4vhm+keo7+6+oSHM9mjJCaInMmf+LfQWy74RN/+3fhbd/PsHDhRr/qiHDJn6ttw4WDzI5LhZR7j+yQ8sYrGNL1fvGtAiF9nAnBPxX6faqCF2V5a8N7LgpDyjxu3o5uRp/E4jZPCg9k80dITRE5k+8NzsCGVqlnkH3NRgXXwbt2sz8Hi/5crQg/QPZv4celn5Fy7pHzjJQ7XsGQBpjt5AqE1FAZeX5J79P42kRIwZPQz9OTyNrRP0M98aMjt/C7bfnJlztoLRMhNUXkTL7LLz7lvmI2Ori1xr8P+mmwcITZURXWPvhkPHj1NM9fr21j9qP0qZtzj5z3SLnjFQrpbxa+hyrwHml7s++k75T+/VWjaxMhBdtODm5Ovfzyv2fvmLDhofZmD33Zw9YSEVJTZM7klf6lXeVSd6lZm+Cd+G/9rQX++jdmO6w+2eywhkRIE1ziU7uZ6VM35x7v+o2zMg+QO17jIa2/sdqsx7KCIZ3l38Ct9LudMPbCtzMhNbo2GdLZZp1XOLeso9ktWTuuvnLMUeF39w43u7uER7bZIaSmSH3YcMaIzha+8Ql+79P3Dw9MjCc+9Xqttdkj4Rcbrg3P8ljsnAcn+yeXHTZk/R4pco/1MbMBt/899QA5WxsJ6cjhwwdv6a+rg8/cCoRUX2u2/x1/H262d0MmpEbXJkP6r3/W3O9vU/uZ9V6dvaN/DTn8gTmPXdbKqt8txyFuLgipKaLf/rYBwXcZ5m6dvPWN1b6K/cPP54LnoJrwt6VtLg231TzhIt9siN7Dua8Fi0PTj5CzNT+kpJ5znCsYkrs9ntirT330O36NrU2G5P63LrGt2/ycHed3Tz5i1c2lO67NECE1RfpMru559F8Tv59cfvk+7WJbDb83WL7cvyMPv6w2xP9hvtafqzXulr3rOgx7PliXPqGj93DuraM71Pa5IvMQ2VsbDalqy8HXf+qiY+Z9afWV8b2rO+71kw+ij9vo2lRIbuGZ29e22ePSFbk7uncu2Kt9Ze1Op85THMCWg5DK5sv9lVg0D4RUNoTUkhFS2RBSS0ZIZUNILRkhlQ0htWSEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIFCGkOY+CzQrc4s/y0sf0hwDmpk5RZ/mpQ9ptq0p+WMAQmtsdtH3ISQgByEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIg0ExDWjNjeuktKfnPiRajmYZ0b0W7z9c2FovVfoH9CoidXfKfEy1GMw3pzg4vfb77zOy4L7BfAYefVfKfEy0GIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQaG4hTZoUXm22Ib3R93P+iMHG0dxCGjMmvNpsQ3rOPtzYU0BjCImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQmpeIZ1gsY0jnrjaKRZrF94IblenNlb7m8ffldxlQCe/0CEe3iceq43H6vq0i03ePhar3S02sEv1hbHqySds47dUx/rEfnZbvLXfNV4XX+vciwdOiNUlfswH+77z1qn77Rd7YdGty92w2sc/jMVudO4PXT66Ye246i5L739uwdsPPvWvF/yeK1dED89HY45cFFy3q2ns4M36Z3C5/N/Xr1m93Ll5L7kl89555a31C9a8ui65x4YlOXdZv81TiYXYleFVn9ML/IchpOYUUoVtBvZY/MX3jfessp7f/Sh5eLolVtYkDlNd9rFbGAtWVh1amdipMnug9lf5XR4d2NraDX0hc5/EThWp660T+05v7D8NITWjkFp8RxV5C/mbclzz3I19d/sgPDxBKRXhfhWVlf66Onrs5vnABgysC7bGg41WFe5ZFe7e0drZSHdL1diHXr7vmOppkeNdEeybuk5d3dXIfxtCaj4htdKdsZumSvd6YmHH3C2xrFuJqKr8dYXFnVux6ynB4Rnr19U7t8pv3c7fjFvWeRqzy/3lSL/x9+1eWeR37eOr8Fd9zba0LofbQLut9rpwz4ldks9wJ5mtCq4rkiP5hwyuKq2xAAip+YTU4p+QKp3bP7genlrx9dRC8LMn/hxJH4RKqwrXzHDu7rqVLjzBRwWHKVi5Njxc20UOntUGl7X+GeirE90DyaH2qLb+/rqm4o7KQfE+OzWEe67uNDV1vI9KHffgqpOlg2rI/4/T3EIavMOp3gPlCunDCcHDfbdhk1gq/Zm8sblEBtWp2yvTG9JvaTqmFjpY54rgqepc55bY/MQ9n/BX+wYb/+MXqoIwUx61gf7y06DD6vvceYkhKr7Xzrr6d1Q7Vv28W48d68Yl9/2fixLXqWeexMTCP8dcYuHm/BOTkAhp0+HKEtL9hMRLuxatbC/tHkwOlf3SbufN6aXd5hwSHzYklenDhjF82NAyQ2r5T0kb++PvcdP4+HtzCKnllxQo6heyvfiF7GchpEL4ihBfESoCIRXCl1Y3UYRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCQHMLadKk8GqzDemNvoWODDaq5hZS0mYbEjZRhERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBgJAICQKEREgQICRCggAhERIECImQIEBIhAQBQiIkCBASIUGAkAgJAoRESBAgJEKCACEREgQIiZAgQEiEBAFCIiQIEBIhQYCQCAkChERIECAkQoIAIRESBAiJkCBASIQEAUIiJAgQEiFBoJmGdG9Fu8/XNhaL1X6B/QqInV3ynxMtRjMNac2M6aW3pOQ/J1qMZhoSsGkhJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRDYNEOaY0AzM6fo07z0Ibm5zxZwxMFTSmPEjiUa+MyOJRr4Uru5NAPfYFeUZuApdeeUaODeJ5Ro4P2OLnQm5ppb/FlehpAKGjOmRAP/aECJBp7avUQDP2UrSzPw+zavNAO7DveUaOB+k0s08DfPLNHAAUIqBiFlEFIWQioGIWUQUhZCKgYhZRBSFkIqBiFlEFIWQioGIWUQUhZCKgYhZRBSFkIqBiFlEFIWQioGIWUQUhZCKgYhZRBSlo0Z0qmnlmjgnw4q0cB3bFuigZ+vKtE3ez+seLU0A7suD5Zo4P6/KtHAo84p0cCBjRnSsmUlGvjTxSUaeN0bJRrYLWx2A7++vkQDv72qRAO/v6JEAwc2ZkhAi0FIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAiUPaQV5/aKdxu/+DNWiAa+JfkPC1ze5JHXXljZ77MeSTWwbsZLz9upttexkb9krppy7sCyKS8ct11N7+NeyqxQzTh3YN1BzlbukFbtbSN+PLZV7/cLrlANfK0dPzEws4kDu1f2bpt1vqtmnDewbMbv96kYdvHoWPyJ1ArVlPMGVk352TbxUZNOjFX/K7VCNeO8gWUHOUe5Q7rGfuYvb7PzCq5QDTzpS/wjN435sHaf+uro+a6acd7Aqhm7s+x6f3mPfS21QjXlvIFVUz6s4jF/ebcNS61QzThvYNlBzlHukPq1Df9G/nZdGwqtUA18ttU3bcSkD85f67LOd9WM8wZWzdidO2Sdv2yo65VaoZpy3sCqKf/g+8Hl+thuqRWqGecNLDvIOcoc0urKgeH1SbawwArVwH5pqVvyXpMGTYue76oZ5w0snbG3OpYaXDrl6MDiKb9uw1MPoZ1xZmD1QU4rc0gLLPH/srvEphdYoRrYHW0/6GTWZ0qThk2Knu+qGecNLJ2x90u7JrkknXJ0YOmUP5y5V5tnksvSGUcHVh/ktDKH9KydFV5fbXcVWKEa2B1qvSf98aJ29tsmjZsQPd9VM84bWDpj5x6N90/9v/KkU44OrJxye7Pj06+6lDPOGlh8kDM2UkhX2d0FVqgGdg/f+bG/nF/TcXWTBg41FlJTZ5w3sHTG7tbqfulPvKRTjg6snPKF4w+qOCj1Qk4546yBtQc5oswh1du3wusf2sMFVqgGTjnWnmlk9yJFz3fVjPMGTlHMuOESG/px+pZwytkDp0gOsnOP1X4l+X+dlB7k6MApohlnlDmkta0S/1vuE+yNAitUA6ecboLfGETPd9WM8wZOEcy4YZyduyFzUzflnIFTJAfZG2GvJBakBzk6cIpqxmnl/vi7f90n/nJ9914FV4gG/vg3fwmvD5Z/uKaace7AwhmfbVdn3ZZNOWdg1ZQX75F4Ajo2/Use0YzzBpaeFlHlDul3drG/vN4uc27VC//OXiEdeEOPupf9igdsz6ZO2aXPd+2McwfWzfiuzO8xtVPOHVg25e3iT/vLBW3arBTPOHdg6WkRVe6Q1h5kwy4dWbGH//PmJRuUvUI78N1VbcZf8o3KNk19LfzoxIkTq7byF++LZ5w3sGrGbjv7Tvg1mInLxFPOG1g15Zmt4sf/YExr+7X6tMgbWHaQc5T9S6ufXNAr3uM7y13qB4usEA8866itY91GLWjquFcmv+Vo9eIZ5w8smrFLDWyvi6ecP7Bqys+P7B3vMOh+Jz8t8gZWzTgHf40CECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQIKTyOs4WNWl74/dZ7NzU3tUXJC6wMRBSWZxv7VeGC8WGdGV9Y3tNCf75yXjXwT9fEe4zZJlbUdvxyn+GF5L5oliEVA5rOlfan8KlIkN6x6Y1ttcUO3DixHNHbm1bprp5zk5PXmCjIKRyuNXOrDgoXCoypPsKhTQpuFp/U21N8l/nnmUTkxe5NqwperooHiGVw6FWf7C9EiwdZwvP6x7f8Xq/uPqqr7Rrs/tVG/zif8d0j3Ue9rRLhDTUgn/Me50N8kveLOcWn75NrPPwzL9onwzJudtt//A90pD0vzd+WnTvb1a8f1j8juiaE23lD3vW7Hhtg19+Y3SXml0mr3XR7ZFJoRiEVAYL7AB3k50XLB5nww685uLudpNzY+2EG24YYROce7NL6wunTOoYfzw3pCdH2yX3fOCW9Gx/0Z+v6BF/LDVgOiTXzxYEIf3rCjv2nhuCi7nRvUfZmIGXzYuuGWfHjJnx8CD7g3/Z2K32rMlft5NcdHtmUigKIZXB+b6bj+o6B6+xjrND/J/29bHeztX1D7eNWO9Osnv80otV++WG5K4MX9qdVvWsv3yz7T6pATMhXejfewWf2kVe2kX2HmdDNmSvGW/f8Iv/saHOnWz/8ItD7aXo9sykUBRCKr3VnWs/dG60/dUFodwWrDrE3nQdur2b2N7QfqvglZY7wJY2HlKnXRcHhvjtCZmQrrPJuSFF9h5vt7rcNQ8Fa2r2dK5jz2Bp4cyl0e3pSaE4hFR6U22Uv5xpg10Q0svBqvH+nc/V1v7Em992wUdzh4X7jbUnGw3pnfQboOeSI2ZCutquywkpuvd4m+Ny18wP7tZ+V7c4nI/L2Z6eFIpDSKV3iP2+vr7+ta4VC4OQ3ghWfdumO/fPo1tbxRH1rt6Ghfud5Vc2FlK97TktYXlyxExIZ/hXhdkhRfceb8GvofLXBCGlHjVn/NSkUBxCKrlX03/gXxSE9GqwbrzNDq7WPHJy5fZrFtvAcMex9lQkpE/TIb1ru+UMmQ5pw9axpTkhRfdOZJO/JvGMNDC5Lmf8xKRUP/3mgpBK7jw7+Y7AlKpu6/xJf1+w7tDgWz2hCfaM69Q1fI+0b8XyMKSjbYm/NT/zHqlz9QfB9vfSQ6ZDmmzHuNz3SJG9k9nkr/EhuY5dgkd99dcv548fTApFIaRSW71FdfIMHWH3Bh9/+9P39fgubnb38LsOE+x5d7Ld7ZeeqxiU+NTuDHvcBR/1+ZtXhVvOsEv95XtbDU+NmfqF7C9btX8zL6TI3sls8tcEIY0LP/cYaS9EtkcmhaIQUqlNtbHJpUdtqPumDTn8hsm9/Dm8Ztf4KddfN7bywAb3drfWP7z1x53bvpgI6R+297THzh/U/jDn7rR9Jz/tlvSsPPmPV/SsnpEaM/yK0PdG9bCtn3R5IUX2TmaTvyYI6a2u1ROu/rp9K7o9MikUhZBK7WCbm1rcvWrRMbbsnG7xnW/xt947Z/vW7Xb/ycd+8c2x3Vp1GRl89SH8itAtu9Z2Pe2j7gc5t3ZEXY+7nFt8xjaxrkc9mR4z/NKqdd736vAlWU5Ikb2T2eSvCUJyr4/qUrPL1WuytkcmhWIQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAv8PsZJtvg0Q7fEAAAAASUVORK5CYII=",
      "text/plain": [
       "Plot with title “Boxplot of Differences”"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "boxplot(abs(as.numeric(unlist(corrected_expr)) - as.numeric(unlist(fed_expression))),\n",
    "        main = \"Boxplot of Differences\", xlab = \"Absolute Differences\",\n",
    "        # add values to the plot - for the mean and the median\n",
    "        horizontal = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
